control-plane/subcommand/sync-catalog/command.go

```diff
@@ -436,6 +459,73 @@ func (c *Command) Run(args []string) int {
 	}
 }
 
+// remove all k8s services from Consul.
+func (c *Command) removeAllK8SServicesFromConsulNode(consulClient *api.Client) error {
+	node, _, err := consulClient.Catalog().NodeServiceList(c.flagPurgeK8SServicesFromNode, &api.QueryOptions{Filter: c.flagFilter})
+	if err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	services := node.Services
+	errChan := make(chan error, 1)
+	batchSize := 300
+	maxRetries := 2
+	retryDelay := 200 * time.Millisecond
+
+	// Ask for user confirmation before purging services
+	for {
+		c.UI.Info(fmt.Sprintf("Are you sure you want to delete %v K8S services from %v? (y/n): ", len(services), c.flagPurgeK8SServicesFromNode))
+		var input string
+		fmt.Scanln(&input)
+		if input = strings.ToLower(input); input == "y" {
+			break
+		} else if input == "n" {
+			return nil
+		} else {
+			c.UI.Info("Invalid input. Please enter 'y' or 'n'.")
+		}
+	}
+
+	for i := 0; i < len(services); i += batchSize {
+		end := i + batchSize
+		if end > len(services) {
+			end = len(services)
+		}
+
+		wg.Add(1)
+		go func(batch []*api.AgentService) {
+			defer wg.Done()
+
+			for _, service := range batch {
+				var b backoff.BackOff = backoff.NewConstantBackOff(retryDelay)
+				b = backoff.WithMaxRetries(b, uint64(maxRetries))
+				err := backoff.Retry(func() error {
+					_, err := consulClient.Catalog().Deregister(&api.CatalogDeregistration{
+						Node:      c.flagPurgeK8SServicesFromNode,
+						ServiceID: service.ID,
+					}, nil)
+					return err
+				}, b)
+				if err != nil {
+					if len(errChan) == 0 {
+						errChan <- err
+					}
+				}
+			}
+			c.UI.Info(fmt.Sprintf("Processed %v K8S services from %v", len(batch), c.flagPurgeK8SServicesFromNode))
+		}(services[i:end])
+		wg.Wait()
+	}
+
+	close(errChan)
+	if err = <-errChan; err != nil {
+		return err
+	}
+	c.UI.Info("All K8S services were deregistered from Consul")
+	return nil
```

Comment:
We discussed a bit outside of the PR that we _want_ to execute these requests in serial. Given that, I'm not sure why the `WaitGroup` and goroutine and error channel are necessary here. Why not just?

```suggestion
	var firstErr error
	for i := 0; i < len(services); i += batchSize {
		end := i + batchSize
		if end > len(services) {
			end = len(services)
		}
		
		batch := services[i:end]
		for _, service := range batch {
			var b backoff.BackOff = backoff.NewConstantBackOff(retryDelay)
			b = backoff.WithMaxRetries(b, uint64(maxRetries))
			err := backoff.Retry(func() error {
				_, err := consulClient.Catalog().Deregister(&api.CatalogDeregistration{
					Node:      c.flagPurgeK8SServicesFromNode,
					ServiceID: service.ID,
				}, nil)
				return err
			}, b)
			if err != nil {
				if firstErr == nil {
					firstErr = err
				}
			}
		}
		c.UI.Info(fmt.Sprintf("Processed %v K8S services from %v", len(batch), c.flagPurgeK8SServicesFromNode))
	}
	
	if firstErr != nil {
		return err
	}

	c.UI.Info("All K8s services were deregistered from Consul")
	return nil
```

Reply:
Updated the code to execute the batch in sequence but execute the tasks inside batch concurrently